cdap-master/src/main/java/co/cask/cdap/data/tools/RemoteDSService.java

```diff
@@ -0,0 +1,145 @@
+import co.cask.cdap.metrics.guice.MetricsClientRuntimeModule;
+import co.cask.cdap.notifications.feeds.guice.NotificationFeedServiceRuntimeModule;
+import co.cask.cdap.notifications.guice.NotificationServiceRuntimeModule;
+import co.cask.cdap.proto.Id;
+import co.cask.cdap.store.guice.NamespaceStoreModule;
+import com.google.inject.Guice;
+import com.google.inject.Inject;
+import com.google.inject.Injector;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.twill.zookeeper.ZKClientService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Provides a {@link DatasetService} which uses {@link RemoteDatasetFramework}.
+ * This is used to independently start a remote dataset service during upgrade.
+ */
+public class RemoteDSService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(UpgradeTool.class);
+
+  private final DatasetService datasetService;
+  private final ZKClientService zkClientService;
+  private final DatasetFramework datasetFramework;
+  private final DatasetOpExecutorService datasetOpExecutorService;
+
+  @Inject
+  RemoteDSService(CConfiguration cConf, Configuration hConf) {
+    Injector injector = initInjector(cConf, hConf);
+    this.datasetService = injector.getInstance(DatasetService.class);
+    this.zkClientService = injector.getInstance(ZKClientService.class);
+    this.datasetFramework = injector.getInstance(DatasetFramework.class);
+    this.datasetOpExecutorService = injector.getInstance(DatasetOpExecutorService.class);
+
+  }
+
+  public DatasetFramework getDSFramework() {
+    return datasetFramework;
+  }
+
+  /**
+   * Start the {@link DatasetService} with its dependent services and wait for the dataset service to come up
+   */
+  public void startAndWait() throws Exception {
+    if (!zkClientService.isRunning()) {
+      zkClientService.startAndWait();
+    }
+    datasetOpExecutorService.startAndWait();
+    datasetService.startAndWait();
+
+    // wait 5 minutes for DatasetService to start up
+    Tasks.waitFor(true, new Callable<Boolean>() {
+      @Override
+      public Boolean call() throws Exception {
+        try {
+          getDSFramework().getInstances(Id.Namespace.DEFAULT);
+          return true;
+        } catch (ServiceUnavailableException sue) {
+          return false;
+        }
+      }
+    }, 5, TimeUnit.MINUTES, 10, TimeUnit.SECONDS);
+  }
+
+  /**
+   * Stop the running dataset service
+   */
+  public void stop() {
+    try {
+      datasetService.stopAndWait();
+      datasetOpExecutorService.startAndWait();
+      zkClientService.stopAndWait();
+    } catch (Throwable e) {
+      LOG.error("Exception while trying to stop services running for {}", RemoteDSService.class.getSimpleName(), e);
+      Runtime.getRuntime().halt(1);
+    }
+  }
+
+  private static Injector initInjector(CConfiguration cConf, Configuration hConf) {
+    return Guice.createInjector(
```

Comment:
Are all these modules needed? For example, we shouldn't need ``NamespaceClientRuntimeModule`` after #5109 right?